ENH: Add single-sided p-values to t-tests

DominicChm opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
Part of work being done for gh-12506.

#### What does this implement/fix?
This enhancement implements the "alternative" parameter in SciPy's t-tests to allow the selection and return of single-sided p-values. 

#### Additional information
•Implementation closely follows the method found in the brunnermunzel test with modifications made to support arrays. •Selection is done by setting the alternative parameter to one of three strings ("two-sided", "less", or "greater") with the default value being "two-sided" for backwards-compatibility.
•I'm still in the process of writing unit tests. They will be implemented as a method in the "TestStudentTest" class in test_stats.py, if it's the best place.

mdhaber commented on scipy/stats/stats.py:
```diff
@@ -5329,26 +5338,37 @@ def ttest_1samp(a, popmean, axis=0, nan_policy='propagate'):
 
 
-def _ttest_finish(df, t):
+def _ttest_finish(df, t, alternative):
```
Just checking - how did you ensure that you changed all calls to `_ttest_finish`?

DominicChm replied:
I assumed it's a function private to stats.py (because of the _), so I just found all calling functions in stats.py and ensured they could pass through an alternative parameter.

mdhaber commented:
![image](https://user-images.githubusercontent.com/6570539/88234134-62b75c80-cc2d-11ea-8a51-7e4147f7c5fa.png)
At the right, click on "Details" to see what went wrong in the continuous integration tests. The first of these (ARM - Test Group 1) doesn't look like it could be caused by this PR. The ARM builds fail intermittently, so we don't really need to be concerned unless it's happening consistently with this PR and not others. Click on the second (Refguide-check, Latest NumPy) and scroll to the bottom. These failures look like they were caused by [doctests](https://docs.python.org/3/library/doctest.html), which are intolerant of even the most minute changes in output. Just find where those doctests are written and update the numbers; they should go away.

DominicChm commented on scipy/stats/stats.py:
```diff
@@ -7141,8 +7188,7 @@ def kruskal(*args, **kwargs):
        The p-value for the test using the assumption that H has a chi
-       square distribution. The p-value returned is the survival function of
-       the chi square distribution evaluated at H.
+       square distribution.
```
Looks like I might have forgotten to pull before commiting...? Either way, unintentional. Is there any easy way to revert the change or should I just re-commit the doc?

mdhaber replied:
Yeah I think just add it back. You can either do that manually on your own computer, but I want to suggest another way just so you get familiar with using GitHub.

You should see a button that will let you accept this suggestion:
```suggestion
       square distribution. The p-value returned is the survival function of
       the chi square distribution evaluated at H.
```
Then on your machine:
```
git fetch origin
git status
```
Will get the latest information from your remote (`origin`) and tell you the `status` of your local branch relative to the one on the remote. It will tell you that the remote branch is one commit ahead. It might also tell you you have local changes, in which case you should commit (or stash) them before you rebase on your remote branch:
```
git rebase origin/ttest-single-tailed-pvalues
```
(Which is similar to "merging" the changes, but [works a bit differently](https://www.atlassian.com/git/tutorials/merging-vs-rebasing).) 
If you were to accept these changes online without rebasing (or merging), the next time you try to push to the remote it would get an error like [this](https://happygitwithr.com/push-rejected.html).

DominicChm commented on scipy/stats/stats.py:
```diff
@@ -5329,26 +5338,37 @@ def ttest_1samp(a, popmean, axis=0, nan_policy='propagate'):
+    prob = distributions.t.sf(t, df)
+
+    if alternative == "greater":
+        prob = 1 - prob
```
That makes sense. When I was implementing this, I did it under the assumption that `popmean` was being tested against `a`, exactly the opposite of what it should be. I'll push a commit soon to fix it.

mdhaber replied:
OK, please also add a unit test that confirms the p-value `p` relative to the two tailed p-value `p2` (e.g. `np.allclose(p, p2/2)` or  `np.allclose(p, 1-p2/2)`) . You can either find an [example online](https://stats.idre.ucla.edu/other/mult-pkg/faq/general/faq-what-are-the-differences-between-one-tailed-and-two-tailed-tests/), or you can create data so you know what the answer will be.

mdhaber commented on scipy/stats/stats.py:
```diff
@@ -346,6 +346,26 @@ def _broadcast_shapes_with_dropped_axis(a, b, axis):
 
+def _convert_symmetric_p_value(pvalue, root_func, alternative):
+    """
+    Convert a one-sided, symmetric P-value to a one or two-sided p-value. root_func
```
```suggestion
    Convert a one-sided p-value from a symmetric distribution to a one or two-sided p-value.
```
The first line of a docstring should fit on just one line and shouldn't refer to parameters. See https://numpy.org/devdocs/docs/howto_document.html#docstring-standard

mdhaber replied:
Please let me resolve comments so I can confirm that they are actually resolved.

mdhaber commented on scipy/stats/stats.py:
```diff
@@ -346,6 +346,26 @@ def _broadcast_shapes_with_dropped_axis(a, b, axis):
+    Convert a one-sided, symmetric P-value to a one or two-sided p-value. root_func
+    describes the source distribution function ('sf' or 'cdf')
+    """
+    if alternative not in ["two-sided", "greater", "less"]:
```
This is a private function, so only developers will use it directly, and so you don't have to do input validation.

If you were to do input validation:
```suggestion
    if alternative not in {"two-sided", "greater", "less"}:
```
Lookup is O(1) on average because they're implemented using a hash table; lookup in a list is O(n) because they're basically fancy arrays. 
Of course it doesn't matter in practice, just a matter of principle and style.

mdhaber commented on scipy/stats/stats.py:
```diff
@@ -346,6 +346,26 @@ def _broadcast_shapes_with_dropped_axis(a, b, axis):
+        raise ValueError("root_func should be 'sf' or 'cdf'")
+
+    op_type = alternative + "_" + root_func
+    if op_type == "greater_sf" or op_type == "less_cdf":
```
String operations aren't very efficient in Python, but more importantly, I manipulating strings seems a bit indirect in Python. 

Something more familiar to other developers might be:
```suggestion
    if (alternative, root_func) in {("greater", "sf"), ("less", "cdf")}
```

But I was actually going to suggest some renaming that I think would simplify the logic. It seems to me that regardless of whether the function used to calculate the incoming one-sided *p*-value is `sf` or `1-cdf`, either way it's still a `'greater'` *p*-value. Likewise, whether it's implemented as `cdf` or `1-sf` or something else entirely (because remember, there are some tests for which they don't even call a stats `cdf` or `sf`), it's still a `'less'` *p*-value. So maybe if `in_type` is `'less'` or `'greater'` and `out_type` is `'less'`, `'greater'`, `'two-sided'`, then:
```
if out_type == "two-sided":
    return 2 * np.minimum(pvalue, 1 - pvalue)  # although if pvalues are never arrays, maybe just use `min`?
elif out_type != in_type:
    return 1 - pvalue
else:
    return pvalue
```

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -2469,6 +2471,14 @@ def test_onesample(self):
+    def test_1samp_alternative(self):
+        assert_raises(ValueError, stats.ttest_1samp, self.X1, 0, alternative="error")
+
+        t, p = stats.ttest_1samp(self.X1, 1, alternative="less")
```
If you're going to do just one test, you should use `self.X2` instead of `self.X1` because the latter has a p-value of 1/2 with either alternative, so you could have the logic reversed and not catch it.

mdhaber replied:
Never mind, I see that you were using a hypthesized mean `mu=1`, not `mu=0`. This is fine.

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -2424,6 +2424,8 @@ class TestStudentTest(object):
     P1_2 = 0.0741799
     T2_0 = 1.732051
     P2_0 = 0.2254033
+    P1_1_l = P1_1 / 2
```
Looks backwards.

mdhaber replied:
It's not backwards. Never mind.

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3319,6 +3329,14 @@ def test_ttest_rel():
+    assert_raises(ValueError, stats.ttest_rel, rvs1, rvs2, alternative="error")
+
+    t,p = stats.ttest_rel(rvs1, rvs2, axis=0, alternative="less")
+    assert_array_almost_equal([t,p],(tr,1 - pr/2))
```
Note above:
```
    assert_array_almost_equal(np.abs(t), tr)
    assert_array_almost_equal(np.abs(p), pr)
```
This is pretty standard in our tests. Please separate these.
If you were to combine them:
```suggestion
    assert_array_almost_equal((t,p),(tr,1 - pr/2))
```

mdhaber replied:
I'll look at the rest later.

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -2469,6 +2471,14 @@ def test_onesample(self):
+        assert_array_almost_equal(p, self.P1_1_l)
+
+        t, p = stats.ttest_1samp(self.X1, 1, alternative="greater")
+        assert_array_almost_equal(p, self.P1_1_g)
```
Also, I haven't looked at the two-sided tests - does it have any for when the input data is a 2d array? I wouldn't say you have to fix it if not, but you do need to confirm that what you wrote works for 2d arrays.
See the [documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html).
So you'd need to test something like:
```
ttest_1samp([[0, 1, 2], [-1, 0, 1], [-2, -1, 0]], 0, axis=1, alternative='greater')
```
and separately 
```
ttest_1samp([[0, 1, 2], [0, 1, 2]], 0, axis=None, alternative='greater')
```
because the latter tests over the whole array.

Weird that the documentation doesn't mention the default... I would suggest adding that information.

DominicChm replied:
Yep, it looks like it does (in test_ttest_1samp_new() at line 3717 - rvn1 is a multi-dimensional array). I've also tested my implementation and it seems to work fine.

mdhaber replied:
OK, please add to `test_ttest_1samp_new` a test of your `alternative` keyword as suggested above. Also, please add to the `ttest_1samp` documentation what the default axis is.

mdhaber replied:
For example:
```
axis : int or None, optional
        Axis along which to compute test; default is ___. If None, compute over the whole
        array `a`.
```
I haven't checked what the default is.

It's not your fault that it's not already documentated, but it would be nice to fix it while we're working on this.

WarrenWeckesser commented:
Just a quick note: I haven't given the code a careful review, but I did install this PR and compare it to R's `t.test`, including the various combinations of the `t.test` parameters `alternative`, `equal_var` and `paired`, and I got agreement between the R code and the results computed using this PR.

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3319,6 +3329,17 @@ def test_ttest_rel():
+    assert_raises(ValueError, stats.ttest_rel, rvs1, rvs2, alternative="error")
+
+    t,p = stats.ttest_rel(rvs1, rvs2, axis=0, alternative="less")
+    assert_array_almost_equal(np.abs(p), 1 - pr/2)
```
For these new tests, please get rid of all the absolute value signs. Predict what the sign should be, and include it in the test. We want the test to fail if the sign is incorrect.

WarrenWeckesser commented:
@DominicChm, another formatting change you'll need to make is to ensure that any lines that you add or edit have a length of at most 79 characters.  This is a PEP-8 requirement, and with the recently merged https://github.com/scipy/scipy/pull/12734, this will be checked as part of the CI tests for each pull request.

DominicChm commented:
> @DominicChm, another formatting change you'll need to make is to ensure that any lines that you add or edit have a length of at most 79 characters. This is a PEP-8 requirement, and with the recently merged #12734, this will be checked as part of the CI tests for each pull request.

@WarrenWeckesser Got it, I'll get right on making the necessary changes. As far as code style when breaking up longer statements, is the following acceptable? Or is there a more elegant way to force the 79 character limit than splitting lines with a '\\'?
```
...
        raise ValueError("alternative should be 'less', 'greater' or 'two-sided'")
        # becomes:
        raise ValueError("alternative should be "
                         "'less', 'greater' or 'two-sided'")

...
    elif (alternative, root_tail) in {("greater", "less"), ("less", "greater")}:
    # becomes:
    elif (alternative, root_tail) in \
            {("greater", "less"), ("less", "greater")}:
```

mdhaber commented:
> is there a more elegant way to force the 79 character limit 

I've been told that it is preferable to use natural groupings where possible:
```
elif (alternative, root_tail) in {("greater", "less"), 
                                  ("less", "greater")}:
```
Parentheses, square brackets, and curly braces all let you continue on new lines without explicit breaks.

DominicChm commented:
> I've been told that it is preferable to use natural groupings where possible:
> 
> ```
> elif (alternative, root_tail) in {("greater", "less"), 
>                                   ("less", "greater")}:
> ```
Got it, will push a commit soon fixing these.

DominicChm commented:
@WarrenWeckesser Looks like all tests are passing, is this ready to merge?

WarrenWeckesser commented:
@DominicChm, I'm taking a closer look at the code this afternoon.  More soon...

WarrenWeckesser commented on scipy/stats/stats.py:
```diff
@@ -5303,13 +5335,15 @@ def ttest_1samp(a, popmean, axis=0, nan_policy='propagate'):
     Examples using axis and non-scalar dimension for population mean.
 
     >>> stats.ttest_1samp(rvs,[5.0,0.0])
-    (array([-0.68014479,  4.11038784]), array([  4.99613833e-01,   1.49986458e-04]))
+    (array([-0.68014479,  4.11038784]), array([  4.99613833e-01,
+    1.49986458e-04]))
```
I guess a simple linebreak like this works, but it looks a bit messy in the later examples.  How about breaking it up, something like this:
```
>>> result = stats.ttest_1samp(rvs,[5.0,0.0])
>>> result.statistic
array([-0.68014479,  4.11038784])
>>> result.pvalue
array([  4.99613833e-01,   1.49986458e-04])
```

WarrenWeckesser commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3314,11 +3327,23 @@ def test_ttest_rel():
 
-    t,p = stats.ttest_rel(np.rollaxis(rvs1_3D,2), np.rollaxis(rvs2_3D,2), axis=2)
+    t, p = stats.ttest_rel(np.rollaxis(rvs1_3D, 2), np.rollaxis(rvs2_3D, 2),
+                          axis=2)
```
Indent one more space to align with the first argument above:
```suggestion
                           axis=2)
```

WarrenWeckesser commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3462,11 +3487,37 @@ def test_ttest_ind():
 
-    t,p = stats.ttest_ind(np.rollaxis(rvs1_3D,2), np.rollaxis(rvs2_3D,2), axis=2)
+    t, p = stats.ttest_ind(np.rollaxis(rvs1_3D, 2), np.rollaxis(rvs2_3D, 2),
+                          axis=2)
```
```suggestion
                           axis=2)
```

WarrenWeckesser commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3493,7 +3544,6 @@ def test_ttest_ind():
         assert_equal(stats.ttest_ind(anan, np.zeros((2, 2))),
                      ([0, np.nan], [1, np.nan]))
 
-
```
Don't delete this line--module-level `def`s should have two blank line before them (PEP 8).

WarrenWeckesser commented on scipy/stats/tests/test_stats.py:
```diff
@@ -2469,6 +2471,17 @@ def test_onesample(self):
+                      alternative="error")
+
+        t, p = stats.ttest_1samp(self.X1, 1, alternative="less")
+        assert_array_almost_equal(p, self.P1_1_l)
```
Here and elsewhere, use `assert_allclose` instead of `assert_array_almost_equal` in the tests that you add.  While we have a lot of uses of `assert_array_almost_equal` in the existing code, in new code we should be using `assert_allclose`.  (I think we're all still working on adapting to this preferred API--I've used `assert_array_almost_equal` in recent PRs myself.)

mdhaber commented:
@WarrenWeckesser I think the plan was to use `_convert_symmetric_p_value` in all tests with a symmetric *p*-value, not just here. You would prefer for the same logic to be written out for each test?

mdhaber commented on scipy/stats/stats.py:
```diff
@@ -346,6 +346,28 @@ def _broadcast_shapes_with_dropped_axis(a, b, axis):
+    if root_tail not in {"greater", "less"}:
+        raise ValueError("root_func should be 'less' or 'greater'")
+
+    if (alternative, root_tail) in {("greater", "greater"),
```
This is if `alternative == root_tail`.

You can simplify this a lot if you start by checking for alternative two-sided, next check for the equality above, and end with `else`.

mdhaber replied:
See original comment about this [here](https://github.com/scipy/scipy/pull/12597#pullrequestreview-457179588)

WarrenWeckesser commented:
@mdhaber, yeah, it's just a few lines of straightforward code.   It also avoid the loss of precision that occurs in the expression `1 - p` when `p` is very small.

mdhaber commented:
@DominicChm - @WarrenWeckesser and I have discussed this. As I wrote originally:
> And maybe someone else will have something better and ask you to change it again : )

So go ahead and change as Warren suggests. Sometimes you end up right where you started, and that's not necessarily a bad thing. It can be tough to identify the best solution without seeing the alternatives. And hey - after copying and pasting this in a few other PRs, we might even decide to revisit this again!

mdhaber commented:
Maybe eventually we'll want a function with a signature:
```
def statistic_to_p(statistic, distribution, alternative):
```
That would evaluate `distribution`'s `cdf` or `sf`, depending on `alternative` at the value of `statistic`.

But yeah, for now, let's just keep it simple.

WarrenWeckesser commented:
@DominicChm, the latest changes look good, thanks.

I was running through a bunch of different permutations of inputs, and I realized there is one big problem remaining.  If `nan_policy` is `'omit'` and the input contains `nan`, the functions pass the arguments off to the `mstats` versions, and those versions don't handle `alternative`.  This means an example such as

```
In [71]: x = np.array([1, 1, 3, 5, 8, 5, 4, -3, np.nan])

In [72]: ttest_1samp(x, 1, nan_policy='omit', alternative='less')          
Out[72]: Ttest_1sampResult(statistic=1.694637091472647, pvalue=0.13396903280382275)
```

gives an incorrect result; the correct result is

```
In [74]: ttest_1samp(x[:-1], 1, alternative='less')
Out[74]: Ttest_1sampResult(statistic=1.694637091472647, pvalue=0.9330154835980886)
```

I've worked on removing the use of the `mstats` functions for handling `nan_policy='omit'` in other pull requests, so I know from experience it gets a little messy, especially when we also have an `axis` argument like we do here.  Eventually I think we'll end up with a collection of tools that will make this much easier, but we're not there yet.

@mdhaber (and anyone else): what do you think of temporarily disallowing the combination of `nan_policy='omit'` and `alternative` being either `'less'` or `'greater'` (e.g. raise an exception in that case), with the plan to fix this before 1.6 is released?

We could also work on adding `alternative` to the `mstats` functions, but ultimately, calling the `mstats` versions when `nan_policy='omit'` still isn't what we want, so it doesn't really solve the immediate problem.

WarrenWeckesser commented:
> We could also work on adding alternative to the mstats functions, but ultimately, calling the mstats versions when nan_policy='omit' still isn't what we want, so it doesn't really solve the immediate problem.

On the other hand, maybe if we use the new `_ttest_finish` function in the `mstats` versions of `ttest_*`, it will be easy to add `alternative` to these functions.  Then we could--for now--continue to call the `mstats` version, and leave that clean up for another time.

WarrenWeckesser commented:
P.S. The idea to create `_evaluate_distribution_p` to handle the alternative p-values more generically looks good, but lets wait for the next PR to add that generalization. (The name should also have `symmetric` in it somewhere.)

mdhaber commented:
I don't have any experience with mstats so please make that call @WarrenWeckesser.

`_evaluate_distribution_p` doesn't seem to rely on the distribution being symmetric ,does it? I think it just needs a distribution for which the different alternatives are meaningful.

rlucas7 commented:
Hmm, IIUC it seems either way is a little messy. Is this right?

I’d prefer this to throwing an exception given that we’d want the other behavior in before 1.6 release and the release will likely be upon us sooner than we now think.

Does the finishing function work as in on the masked arrays? I haven’t seen that used yet but I might’ve missed a PR or 3, this summer’s been busier than I anticipated. 

-L

> On Aug 25, 2020, at 9:56 PM, Warren Weckesser <notifications@github.com> wrote:
> 
> ﻿
> We could also work on adding alternative to the mstats functions, but ultimately, calling the mstats versions when nan_policy='omit' still isn't what we want, so it doesn't really solve the immediate problem.
> 
> On the other hand, maybe if we use the new _ttest_finish function in the mstats versions of ttest_*, it will be easy to add alternative to these functions. Then we could--for now--continue to call the mstats version, and leave that clean up for another time.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

DominicChm commented:
@WarrenWeckesser Ah, I completely didn't notice that little check. At the moment my plan was to leave the masked array tests for later because they seem to use unusual methods to generate their P-values ( masked `ttest_1samp` for example calculates its pvalue using `prob = special.betainc(0.5*df, 0.5, df/(df + t*t))` ).  Is this is because the CDF and SF distribution functions don't support masked arrays? It's not clear to me because other functions like the masked `skewtest` seem to have no problem using the standard (and easily changeable) `2 * distributions.norm.sf(np.abs(...))` method. 

@rlucas7 Doing some quick testing seems to indicate that the T sf and cdf throw an invalid value error when passed a NaN inside a masked array, but the current masked implementation seems to expect and ignore these errors given that part of the calculation is wrapped with `with np.errstate(divide='ignore', invalid='ignore'):`. 

To test a bit more, I removed the call to the masked version inside the standard function and adjusted how `n` is calculated in the normal ttest_1samp to handle masked arrays, finding that it seems to work correctly with a (very simple) test array = [1, 2, float("nan"), 4, 5] 
```
def ttest_1samp(a, popmean, axis=0, nan_policy='propagate',
                alternative="two-sided"):
    
    a, axis = _chk_asarray(a, axis)

    contains_nan, nan_policy = _contains_nan(a, nan_policy)

    if contains_nan and nan_policy == 'omit':
        a = ma.masked_invalid(a)
        n = a.count(axis) #+
        #return mstats_basic.ttest_1samp(a, popmean, axis)
    else :
        n = a.shape[axis] #+

    df = n - 1

    d = np.mean(a, axis) - popmean
    v = np.var(a, axis, ddof=1)
    denom = np.sqrt(v / n)

    with np.errstate(divide='ignore', invalid='ignore'):
        t = np.divide(d, denom)
    t, prob = _ttest_finish(df, t, alternative)

    return Ttest_1sampResult(t, prob)
```

There is certainly more nuance to it (proxying the masked version to it fails some tests), but could a possible solution be to simply merge the masked and standard versions of the function?

WarrenWeckesser commented:
@DominicChm,   sorry for the long gap since my last comment!  Your suggested modification to `ttest_1samp` works for the examples that I've tried.  For example, this session is run with your suggested changes:

```
In [99]: a
Out[99]: 
array([[ 8.,  7.,  1.,  0.,  6.,  4.,  7.,  4.],
       [ 7.,  6.,  9.,  0.,  1., nan, nan, nan],
       [ 6.,  3.,  5.,  0.,  6.,  5.,  6., nan]])

In [100]: stat, p = ttest_1samp(a, 5, axis=1, nan_policy='omit', alternative='less')                                                         

In [101]: stat
Out[101]: 
masked_array(data=[-0.36266239, -0.22866478, -0.67936622],
             mask=False,
       fill_value=1e+20)

In [102]: p
Out[102]: array([0.36377991, 0.41517215, 0.26111822])

In [103]: for b in a: 
     ...:     print(ttest_1samp(b[~np.isnan(b)], 5, alternative='less')) 
     ...:                                                                                                                                    
Ttest_1sampResult(statistic=-0.36266238893870795, pvalue=0.3637799139221004)
Ttest_1sampResult(statistic=-0.228664780190012, pvalue=0.415172151852777)
Ttest_1sampResult(statistic=-0.6793662204867571, pvalue=0.26111822031871)
```

It doesn't eliminate the behavior that results in the surprising use of masked arrays for the return values when `nan_policy` is `'omit'`, but that is existing behavior that we don't have to change in this PR.   Can you go ahead and implement that, and include some unit tests?  Something like the above example could be used as a test--that is, compare the result for an input that contains `nan` to the result obtained when the `nan`s are removed before calling `ttest_1samp`.

Have you tried something similar for `ttest_ind` and `ttest_rel`?
Then

DominicChm commented:
@WarrenWeckesser I haven't yet tried it with _rel or _ind. I should also mention that I tried to entirely replace the masked version of this test with a call to my above test implementation and a few of the masked tests failed. I'll see if I can work around it in some way, and try to add this functionality to the other tests.

WarrenWeckesser reviewed (changes requested):
Other than the request to replace a couple more uses of `assert_array_almost_equal` with `assert_allclose`, my suggestions are all cosmetic.  I'm OK with not handling the one-sided alternatives for inputs that would have previously been passed to the mstats version.   We definitely want to include that, but I think it can be done in a follow-up PR.  As it is, this is a very nice enhancement to the t-test functions.

WarrenWeckesser commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3462,11 +3487,37 @@ def test_ttest_ind():
+    # ttest_ind
+    t, p = stats.ttest_ind(rvs1_2D.T, rvs2_2D.T, axis=0, alternative="less")
+    args = _desc_stats(rvs1_2D.T, rvs2_2D.T)
+    assert_array_almost_equal(
```
```suggestion
    assert_allclose(
```

WarrenWeckesser commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3462,11 +3487,37 @@ def test_ttest_ind():
+
+    t, p = stats.ttest_ind(rvs1_2D.T, rvs2_2D.T, axis=0, alternative="greater")
+    args = _desc_stats(rvs1_2D.T, rvs2_2D.T)
+    assert_array_almost_equal(
```
```suggestion
    assert_allclose(
```

WarrenWeckesser commented:
@DominicChm, all my suggestions were minor changes, so I've committed them.  Once the tests finish, I think we can merge this.

WarrenWeckesser commented:
@DominicChm, we'll need the same change for `ttest_rel` and `ttest_ind` that you made for `ttest_1samp` to avoid dropping down to masked arrays when `alternative` is not `two-sided`.  For example,
```
In [63]: xn                                                                     
Out[63]: array([-10. ,   1. ,   2. ,   2.5,   9. ,   nan])

In [64]: yn                                                                     
Out[64]: array([-13.,  -4.,   0.,   1.,   3.,  nan])
```
For comparison, here's the default two-sided `ttest_ind` result:
```
In [65]: ttest_ind(xn, yn, nan_policy='omit')                                   
Out[65]: Ttest_indResult(statistic=0.8373780870523444, pvalue=0.4266863609716408)
```
In this case, the one-sided alternatives are ignored, so the result is incorrect:
```
In [66]: ttest_ind(xn, yn, nan_policy='omit', alternative='less')               
Out[66]: Ttest_indResult(statistic=0.8373780870523444, pvalue=0.4266863609716408)

In [67]: ttest_ind(xn, yn, nan_policy='omit', alternative='greater')            
Out[67]: Ttest_indResult(statistic=0.8373780870523444, pvalue=0.4266863609716408)
```
The same is true for `ttest_rel`:
```
In [68]: ttest_rel(xn, yn, nan_policy='omit')                                   
Out[68]: Ttest_relResult(statistic=4.041451884327381, pvalue=0.015585300188062097)

In [69]: ttest_rel(xn, yn, nan_policy='omit', alternative='less')               
Out[69]: Ttest_relResult(statistic=4.041451884327381, pvalue=0.015585300188062097)

In [70]: ttest_rel(xn, yn, nan_policy='omit', alternative='greater')            
Out[70]: Ttest_relResult(statistic=4.041451884327381, pvalue=0.015585300188062097)
```

DominicChm commented:
Thanks for the review, sorry I didn't have time to get to it earlier. I've gone ahead and added the same ValueError to the ttests I didn't get to in the first commit.

WarrenWeckesser commented:
@DominicChm, thanks, I've run this through a bunch of tests of my own and it looks good.

This is a useful improvement as is; handling the masked/nan-containing inputs for the `less` & `greater` alternatives can be done in a follow up PR.

WarrenWeckesser reviewed (approved):
The one test failure is unrelated to this PR.

I'll merge later this weekend if there are no additional comments.

mdhaber commented:
@WarrenWeckesser was this ready?

mdhaber commented:
@WarrenWeckesser would you agree this close gh-10595?

WarrenWeckesser commented:
@mdhaber, yes and yes.  Merged.

WarrenWeckesser commented:
I added a note to the [release notes](https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.6.0).
